libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -1146,6 +1160,65 @@ void WaveClip::Resample(int rate, BasicUI::ProgressDialog *progress)
    }
 }
 
+void WaveClip::ApplyStretchRatio()
+{
+   const auto stretchRatio = GetStretchRatio();
+   if (stretchRatio == 1.0)
+      return;
+
+   Finally Do { [this] {
+      this->mClipStretchRatio = 1.0;
+      this->mRawAudioTempo = this->mProjectTempo;
+   } };
+
+   constexpr auto durationToDiscard = 0.0;
+   constexpr auto blockSize = 1024;
+   const auto numChannels = GetWidth();
+
+   // Let's also pass the hidden part to the stretcher for better result in the
+   // visible part.
+   const auto trimLeftBeforeStretch = mTrimLeft;
+   const auto trimRightBeforeStretch = mTrimRight;
+   SetTrimLeft(0);
+   SetTrimRight(0);
```

Comment:
You make side-effects on the sequence object here, but before all operations have succeeded, and some might fail throwing exceptions because we run out of drive space for the database.

I would rather bullet-proof this.  Add `bool success = false;` and assign it true only at the bottom of the function.

Test `success` in the `Finally` and if it is false, restore trim-left and trim-right to their old values.

Reply:
In other words you are requesting a strong guarantee for this method? I would agree. Coincidentally, not restoring the original trim values is a mistake, even in the case of successful stretching. There's a fix-up commit that does this. I don't see another state-changing method whose side-effect I should revert in case of failure.
It would be useful to have a helper for this safety, like `EffectOutputTracks` making a copy of the track and only replacing the original with it if successful.